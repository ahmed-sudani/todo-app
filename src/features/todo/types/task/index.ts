/* eslint-disable no-unused-vars */
export type TaskType = {
  readonly id: string
  text: string
  completed: boolean
  createdDate: number
  completedDate?: number
}

export enum TASKS_TO_SHOW {
  ALL = 0,
  COMPLETED = 1,
  UNCOMPLETED = 2,
}
