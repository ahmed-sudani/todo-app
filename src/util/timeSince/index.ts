const formateInterval = (label: string, interval: number) => `${Math.floor(interval)} ${label} ago`

export default function timeSince(date: number) {
  const seconds = Math.floor((new Date().getTime() - date) / 1000)
  const times = [
    ['years', 31536000],
    ['months', 2592000],
    ['days', 86400],
    ['hours', 3600],
    ['minutes', 60],
  ] as const

  for (let index = 0; index < times.length; index += 1) {
    const time = times[index]
    const interval = seconds / time[1]
    if (interval > 1) return formateInterval(time[0], interval)
  }

  return formateInterval('seconds', seconds)
}
