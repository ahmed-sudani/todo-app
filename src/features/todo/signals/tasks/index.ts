import { computed, effect, signal } from '@preact/signals'
import { TaskType } from '../../types/task'

export const filteredTasks = signal<TaskType[]>([])
export const tasks = signal<TaskType[]>([])
export const completedTasks = computed<TaskType[]>(() => tasks.value.filter((task) => task.completed) || [])
export const uncompletedTasks = computed<TaskType[]>(() => tasks.value.filter((task) => !task.completed) || [])

effect(() => {
  let initTasks = []
  const strTasks = localStorage.getItem('tasks')
  if (strTasks) initTasks = JSON.parse(strTasks)
  tasks.value = initTasks
})

effect(() => {
  if (tasks.value) localStorage.setItem('tasks', JSON.stringify(tasks.value))
})
