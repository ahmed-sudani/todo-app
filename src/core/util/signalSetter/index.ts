import { Signal } from '@preact/signals'

export const signalSetter = (signal: Signal) => (vl: unknown) => (signal.value = vl)
