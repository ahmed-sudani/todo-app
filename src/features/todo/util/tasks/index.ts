import { completedTasks, tasks, uncompletedTasks } from '../../signals/tasks'
import { TaskType } from '../../types/task'

export const createTask = (task: string) => {
  const newTask: TaskType = {
    completed: false,
    createdDate: Date.now(),
    id: crypto.randomUUID(),
    text: task.trim(),
  }
  if (tasks.value) tasks.value = [...tasks.value, newTask]
  else tasks.value = [newTask]
  return newTask
}

export const updateTask = (id: string, task: Partial<TaskType>) => {
  const newTasks = tasks.value
  const ti = newTasks.findIndex((task) => task.id === id)
  const updatedTask = { ...tasks.value[ti], ...task }
  tasks.value[ti] = updatedTask
  tasks.value = [...newTasks]
  return updatedTask
}

export const completeTaskById = (id: string): TaskType =>
  updateTask(id, {
    completed: true,
    completedDate: Date.now(),
  })

export const completeAllTasks = () => {
  if (!uncompletedTasks.value.length) return
  uncompletedTasks.value.forEach((task) => completeTaskById(task.id))
}

export const deleteAllCompletedTasks = () => {
  if (!completedTasks.value.length) return
  tasks.value = uncompletedTasks.value
}

export const filterTasksByText = ({ tasks, text }: { tasks: TaskType[]; text: string }) => {
  if (!text) return tasks
  return tasks.filter((task) => task.text.search(text) !== -1)
}

export const deleteTaskById = (id: string) => {
  tasks.value = tasks.value.filter((task) => task.id !== id)
}

export const deleteAllTasks = () => {
  tasks.value = []
}
