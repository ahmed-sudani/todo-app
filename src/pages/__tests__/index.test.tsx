/* eslint-disable max-statements */
/* eslint-disable max-lines-per-function */
import { beforeEach, describe, expect, it } from 'vitest'
import { cleanup, fireEvent, render } from '@testing-library/preact'
import { TaskType } from '../../types/task'
import Todo from '..'
import { deleteAllTasks } from '../../util/tasks'

beforeEach(cleanup)
beforeEach(deleteAllTasks)

describe('Todo', () => {
  const taskA: Pick<TaskType, 'text'> = { text: 'a' }

  it('should render successfully', () => {
    const { container } = render(<Todo />)
    expect(container).toMatchSnapshot()
  })
  it('should create task successfully', () => {
    const todoFeature = render(<Todo />)
    const createTaskBtn = todoFeature.getByTestId('feature-todo-create-task-btn')
    const createTaskInp = todoFeature.getByTestId('feature-todo-create-task-input') as HTMLInputElement
    const tasksTable = todoFeature.getByTestId('feature-todo-tasks-table-rows')
    createTaskInp.value = taskA.text
    fireEvent.click(createTaskBtn)
    expect(tasksTable.children.length).toEqual(1)
    const taskAEditInput = todoFeature.getByTestId(
      `feature-todo-tasks-table-row-edit-input-${taskA.text}`
    ) as HTMLInputElement
    expect(taskAEditInput.value).toEqual(taskA.text)
  })
  it('should complete task successfully', () => {
    const todoFeature = render(<Todo />)
    const createTaskBtn = todoFeature.getByTestId('feature-todo-create-task-btn')
    const createTaskInp = todoFeature.getByTestId('feature-todo-create-task-input') as HTMLInputElement
    createTaskInp.value = taskA.text
    fireEvent.click(createTaskBtn)
    const taskARow = todoFeature.getByTestId(`feature-todo-tasks-table-row-${taskA.text}`)
    expect(taskARow.querySelector('.task-status-uncompleted')).toBeInTheDocument()
    expect(taskARow.querySelector('.task-status-completed')).not.toBeInTheDocument()
    const taskCompleteBtn = todoFeature.getByTestId(`feature-todo-tasks-table-row-btn-complete-${taskA.text}`)
    fireEvent.click(taskCompleteBtn)
    expect(taskARow.querySelector('.task-status-uncompleted')).not.toBeInTheDocument()
    expect(taskARow.querySelector('.task-status-completed')).toBeInTheDocument()
  })
  it('should update task text successfully', () => {
    const todoFeature = render(<Todo />)
    const createTaskBtn = todoFeature.getByTestId('feature-todo-create-task-btn')
    const createTaskInp = todoFeature.getByTestId('feature-todo-create-task-input') as HTMLInputElement
    createTaskInp.value = taskA.text
    fireEvent.click(createTaskBtn)
    const taskAEditInput = todoFeature.getByTestId(
      `feature-todo-tasks-table-row-edit-input-${taskA.text}`
    ) as HTMLInputElement
    taskAEditInput.value = taskA.text
    fireEvent.blur(taskAEditInput)
    expect(taskAEditInput.value).toEqual(taskA.text)
  })
  it('should delete task successfully', () => {
    const todoFeature = render(<Todo />)
    const createTaskBtn = todoFeature.getByTestId('feature-todo-create-task-btn')
    const createTaskInp = todoFeature.getByTestId('feature-todo-create-task-input') as HTMLInputElement
    const tasksTable = todoFeature.getByTestId('feature-todo-tasks-table-rows')
    createTaskInp.value = taskA.text
    fireEvent.click(createTaskBtn)
    expect(tasksTable.children.length).toEqual(1)
    const taskDeleteBtn = todoFeature.getByTestId(`feature-todo-tasks-table-row-btn-delete-${taskA.text}`)
    fireEvent.click(taskDeleteBtn)
    expect(tasksTable.children.length).toEqual(0)
  })
  it('should filter tasks by text successfully', () => {
    const todoFeature = render(<Todo />)
    const createTaskBtn = todoFeature.getByTestId('feature-todo-create-task-btn')
    const createTaskInp = todoFeature.getByTestId('feature-todo-create-task-input') as HTMLInputElement
    const tasksTable = todoFeature.getByTestId('feature-todo-tasks-table-rows')
    createTaskInp.value = taskA.text
    fireEvent.click(createTaskBtn)
    const taskB: Pick<TaskType, 'text'> = { text: 'b' }
    createTaskInp.value = taskB.text
    fireEvent.click(createTaskBtn)
    expect(tasksTable.children.length).toEqual(2)
    const filterTasksInp = todoFeature.getByTestId('feature-todo-filter-tasks-input') as HTMLInputElement
    filterTasksInp.value = taskA.text
    fireEvent.input(filterTasksInp)
    expect(tasksTable.children.length).toEqual(1)
  })
  it('should filter tasks by completed tasks successfully', () => {
    const todoFeature = render(<Todo />)
    const taskA: Pick<TaskType, 'text'> = { text: 'a' }
    const createTaskBtn = todoFeature.getByTestId('feature-todo-create-task-btn')
    const createTaskInp = todoFeature.getByTestId('feature-todo-create-task-input') as HTMLInputElement
    const tasksTable = todoFeature.getByTestId('feature-todo-tasks-table-rows')
    const filterBtnCompleted = todoFeature.getByTestId('feature-todo-tasks-filter-btn-completed')
    const filterBtnAll = todoFeature.getByTestId('feature-todo-tasks-filter-btn-all')
    createTaskInp.value = taskA.text
    fireEvent.click(createTaskBtn)
    fireEvent.click(filterBtnCompleted)
    expect(tasksTable.children.length).toEqual(0)
    fireEvent.click(filterBtnAll)
    const taskCompleteBtn = todoFeature.getByText('complete')
    fireEvent.click(taskCompleteBtn)
    fireEvent.click(filterBtnCompleted)
    expect(tasksTable.children.length).toEqual(1)
  })
  it('should filter tasks by uncompleted tasks successfully', () => {
    const todoFeature = render(<Todo />)
    const taskA: Pick<TaskType, 'text'> = { text: 'a' }
    const createTaskBtn = todoFeature.getByTestId('feature-todo-create-task-btn')
    const createTaskInp = todoFeature.getByTestId('feature-todo-create-task-input') as HTMLInputElement
    const tasksTable = todoFeature.getByTestId('feature-todo-tasks-table-rows')
    const filterBtnUncompleted = todoFeature.getByTestId('feature-todo-tasks-filter-btn-uncompleted')
    createTaskInp.value = taskA.text
    fireEvent.click(filterBtnUncompleted)
    expect(tasksTable.children.length).toEqual(0)
    fireEvent.click(createTaskBtn)
    expect(tasksTable.children.length).toEqual(1)
  })
})
