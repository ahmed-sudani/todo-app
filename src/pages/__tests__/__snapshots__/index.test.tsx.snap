// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todo > should render successfully 1`] = `
<div>
  <div
    class="flex flex-col gap-4 mx-2 sm:mx-auto mt-5 max-w-4xl p-2 sm:p-6 border border-gray-200 rounded-lg shadow"
  >
    <h1
      class="text-center text-xl font-semibold text-gray-700"
      data-testid="header"
    >
      Todo App
    </h1>
    <form
      class="flex mt-5"
    >
      <input
        class="input rounded-r-none flex-grow"
        data-testid="feature-todo-create-task-input"
        pattern="([^\\\\s]\\\\s?)+"
        placeholder="Enter a task"
        required=""
        type="text"
      />
      <button
        class="btn btn-rounded btn-primary rounded-l-none"
        data-testid="feature-todo-create-task-btn"
        type="submit"
      >
        Create Task
      </button>
    </form>
    <div
      class="flex items-center gap-5 flex-wrap"
    >
      <div
        class="relative"
      >
        <label
          class="absolute inset-y-0 left-0 flex items-center pl-3"
          for="search"
        >
          <img
            class="w-4 h-4 text-gray-500 dark:text-gray-400"
            src="./icons/search.svg"
          />
        </label>
        <input
          class="input pl-10"
          data-testid="feature-todo-filter-tasks-input"
          id="search"
          placeholder="Search for a task"
          type="text"
        />
      </div>
      <div
        class="flex gap-1"
      >
        <button
          class="btn btn-rounded btn-secondary active"
          data-testid="feature-todo-tasks-filter-btn-all"
        >
          0 all
        </button>
        <button
          class="btn btn-rounded btn-secondary"
          data-testid="feature-todo-tasks-filter-btn-completed"
        >
          0 completed
        </button>
        <button
          class="btn btn-rounded btn-secondary"
          data-testid="feature-todo-tasks-filter-btn-uncompleted"
        >
          0 uncompleted
        </button>
      </div>
    </div>
    <div
      class="flex gap-1 sm:justify-end justify-start items-center"
    >
      <button
        class="btn btn-rounded btn-primary"
      >
        Complete All Tasks
      </button>
      <button
        class="btn btn-rounded btn-danger"
      >
        Delete Completed Tasks
      </button>
    </div>
    <table
      class="text-sm text-left text-gray-500 dark:text-gray-400 border-separate border-spacing-x-0.5 border-spacing-y-1"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Task
          </th>
          <th
            class="hidden sm:table-cell"
            scope="col"
          >
            Created
          </th>
          <th
            class="hidden sm:table-cell"
            scope="col"
          >
            Status
          </th>
          <th
            scope="col"
          >
            Action
          </th>
          <th
            scope="col"
          >
            Delete
          </th>
        </tr>
      </thead>
      <tbody
        data-testid="feature-todo-tasks-table-rows"
      />
    </table>
  </div>
</div>
`;
